### PR TITLE
Apiserver proxy rewrites URL when service returns absolute path with request's host

### LINKED ISSUE #51790: apiserver proxy feature does not rewrite Location header on redirects
**Is this a BUG REPORT or FEATURE REQUEST?**:

> /kind bug

**What happened**:
When accessing a service (e.g., `zipkin`) by using the Apiserver proxy feature and the service returns `302 Found` with `Location` header, such as
```
curl http://159.146.11.73:8080/api/v1/namespaces/default/services/zipkin/proxy/ -vvv
... 
< HTTP/1.1 302 Found
< Location: http://159.146.11.73:8080/zipkin/
...
```
The returned `Location` header does not preserve the `api/v1/namespaces/.../proxy/` path if the service sent a redirect with an absolute path. 

Depending on the cluster, the `Location` header written by the service itself might have been `< Location: http://zipkin:9411/zipkin/` and the apiserver proxy did rewrite the _host_:_port_ to match the correct values. Unfortunately, it did not rewrite the path segment. 

**What you expected to happen**:
I would expect the apiserver to rewrite the path segment of the `Location` response header so that it reflects the fact that it's proxied under the `api/v1/.../proxy/` path:
```
< HTTP/1.1 302 Found
< Location: http://159.146.11.73:8080/api/v1/namespaces/default/services/zipkin/proxy/zipkin/
```

**How to reproduce it (as minimally and precisely as possible)**:
```
export KUBECONFIG=/path/to/your/kubeconfig

# use Helm to install Istio which includes Zipkin, unfortunately, there is no Chart for Zipkin only
helm repo add incubator http://storage.googleapis.com/kubernetes-charts-incubator
helm install --name test incubator/istio

kubectl proxy
curl -vvv http://localhost:8001/api/v1/namespaces/default/services/test-istio-zipkin:http/proxy/
...
< HTTP/1.1 302 Found
< Location: http://localhost:8001/zipkin/
< Content-Type: text/plain; charset=utf-8
...
```
The `/api/v1/namespaces/default/services/test-istio-zipkin:http/proxy` path was not preserved. 

**Anything else we need to know?**:
To work around this issue, a service may return strictly a relative url in the `Location` header such as:
```
< Location: ./zipkin/
```
The feature to rewrite `Location` headers is quite common. For instance, Apache Httpd does it thanks to: http://httpd.apache.org/docs/2.2/mod/mod_proxy.html#proxypassreverse

**Environment**:
- Kubernetes version (use `kubectl version`):
```
% kubectl version
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.5", GitCommit:"17d7182a7ccbb167074be7a87f0a68bd00d58d97", GitTreeState:"clean", BuildDate:"2017-08-31T19:32:12Z", GoVersion:"go1.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.2", GitCommit:"922a86cfcd65915a9b2f69f3f193b8907d741d9c", GitTreeState:"clean", BuildDate:"2017-07-21T08:08:00Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration**: _custom_
- OS (e.g. from /etc/os-release): `Ubuntu 16.04.3 LTS`
- Kernel (e.g. `uname -a`): `4.4.0-89-generic`

### PR DESCRIPTION
**Pull request by roycaihw**
**What this PR does / why we need it**:
When a service responses with an URL using an absolute path and the request's host (e.g. in redirection location), current transport recognizes the URL as a different host and doesn't rewrite the absolute path.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #51790

**Special notes for your reviewer**:

**Release note**:

```release-note
Apiserver proxy rewrites URL when service returns absolute path with request's host.
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go [hunk 1]
```diff
@@ -131,21 +131,27 @@ func (rt *Transport) WrappedRoundTripper() http.RoundTripper {
 
 // rewriteURL rewrites a single URL to go through the proxy, if the URL refers
 // to the same host as sourceURL, which is the page on which the target URL
-// occurred. If any error occurs (e.g. parsing), it returns targetURL.
-func (t *Transport) rewriteURL(targetURL string, sourceURL *url.URL) string {
+// occurred, or if the URL matches the sourceHost. If any error occurs (e.g.
+// parsing), it returns targetURL.
+func (t *Transport) rewriteURL(targetURL string, sourceURL *url.URL, sourceHost string) string {
```
Comment (mbohlool) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go]: nit: It was a little confusing for me to read `sourceURL.host` vs `sourceHost`. It is much clear to call it `sourceReqHost` or `sourceRequestHost`

**Comment by mbohlool** [comment]
just one nit.

/lgtm

**Comment by mbohlool** [comment]
@caesarxuchao for approval.

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go [hunk 2]
```diff
@@ -131,21 +131,27 @@ func (rt *Transport) WrappedRoundTripper() http.RoundTripper {
...
 	url, err := url.Parse(targetURL)
 	if err != nil {
 		return targetURL
 	}
 
-	isDifferentHost := url.Host != "" && url.Host != sourceURL.Host
+	isDifferentHost := url.Host != "" && url.Host != sourceURL.Host && url.Host != sourceHost
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go]: This is difficult to reason about if sourceURL.Host and sourceHost are both non-empty… would it make sense to collapse to a single sourceHost param to this function?

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go]: Agree that the names are almost impenetrable

Comment (roycaihw) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go]: Updated.

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go (hunk 2 above)
Comment (roycaihw) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go]: @liggitt Sorry I didn't get the idea here. I think there are cases where sourceURL.Host and sourceRequestHost are both non-empty, and we need to examine both of them to tell if the targetURL is in a different host.

**Comment by roycaihw** [comment]
@liggitt @mbohlool I added some comments to explain the difference between sourceURL.Host and sourceRequestHost (and why we need both of them). Please take a look :)

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go [hunk 3]
```diff
@@ -131,21 +131,39 @@ func (rt *Transport) WrappedRoundTripper() http.RoundTripper {
...
+// occurred, or if the URL matches the sourceRequestHost. If any error occurs (e.g.
+// parsing), it returns targetURL.
+func (t *Transport) rewriteURL(targetURL string, sourceURL *url.URL, sourceRequestHost string) string {
 	url, err := url.Parse(targetURL)
 	if err != nil {
 		return targetURL
 	}
 
-	isDifferentHost := url.Host != "" && url.Host != sourceURL.Host
+	// Example:
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go]: Thanks. An example's worth a thousand words.

**Review by liggitt — COMMENTED** [review]
/approve

**Pull request merged**